File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/entity/component/TagComponent.java
package net.bestia.zoneserver.entity.component;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * The tag component allows attach simple data to the entity.
 * 
 * @author Thomas Felix
 *
 */
public class TagComponent extends Component {

	private static final long serialVersionUID = 1L;

	public static final String TAG_PERSIST = "persist";

	private final Map<String, Object> data = new HashMap<>();

	public TagComponent(long id, long entityId) {
		super(id, entityId);
		// no op.
	}

	public void clear() {
		data.clear();
	}

	public void add(String key, Object value) {
		
		// Only allow serializable values.
		if(!(value instanceof Serializable)) {
			new IllegalArgumentException("Value object must be serializable.");
		}
		
		data.put(key, value);
	}

	public boolean has(String tag) {
		return data.containsKey(tag);
	}

	@Override
	public String toString() {
		return String.format("TagComponent[%s]", data);
	}

	public <T> Optional<T> get(String key, Class<T> type) {

		if (data.containsKey(key)) {
			final Object obj = data.get(key);

			if (type.isAssignableFrom(type)) {
				return Optional.of(type.cast(obj));
			} else {
				return Optional.empty();
			}

		} else {
			return Optional.empty();
		}
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/entity/component/TagComponent.kt
package net.bestia.zoneserver.entity.component

/**
 * The tag component allows attach simple data to an entity.
 *
 * @author Thomas Felix
 */
data class TagComponent(
    override val entityId: Long,
    val tags: Set<String>
) : Component {

  fun has(tag: String): Boolean {
    return tags.contains(tag)
  }

  companion object {
    /**
     * This tagged entity is a usual bestia mob.
     */
    const val MOB = "mob"

    /**
     * This tagged entity is a usual bestia NPC.
     */
    const val NPC = "npc"

    /**
     * Items lying on the ground which can be picked up by the player.
     */
    const val ITEM = "item"

    /**
     * A natural resource which can be harvested if the needed skills are
     * learned by the player.
     */
    const val RESOURCE = "resource"

    /**
     * Entity is under the control of a player.
     */
    const val PLAYER = "player"
  }
}
